pkg/ovs/openflow/ofctrl_action.go

```diff
@@ -0,0 +1,293 @@
+	}
+	if tableID != LastTableID {
+		if repr != "" {
+			repr += ","
+		}
+		repr += fmt.Sprintf("table=%d", tableID)
+	}
+	if zone > 0 {
+		if repr != "" {
+			repr += ","
+		}
+		repr += fmt.Sprintf("zone=%d", zone)
+	}
+	ct := &ofCTAction{
+		ctBase:  base,
+		builder: a.builder,
+		repr:    repr,
+	}
+	return ct
+}
+
+// ofCTAction is a struct to implement CTAction.
+type ofCTAction struct {
+	ctBase
+	actions    []openflow13.Action
+	builder    *ofFlowBuilder
+	strActions []string
+	repr       string
+}
+
+// LoadToMark is an action to load data into ct_mark.
+func (a *ofCTAction) LoadToMark(value uint32) CTAction {
+	a.strActions = append(a.strActions, fmt.Sprintf("load:0x%x->%s[]", value, NxmFieldCtMark))
+	field, rng, _ := getFieldRange(NxmFieldCtMark)
+	a.load(field, uint64(value), &rng)
+	return a
+}
+
+// LoadToLabelRange is an action to load data into ct_label at specified range.
+func (a *ofCTAction) LoadToLabelRange(value uint64, rng *Range) CTAction {
+	a.strActions = append(a.strActions, fmt.Sprintf("load:0x%x->%s[%d..%d]", value, NxmFieldCtLabel, rng[0], rng[1]))
+	field, _, _ := getFieldRange(NxmFieldCtLabel)
+	a.load(field, value, rng)
+	return a
+}
+
+func (a *ofCTAction) load(field *openflow13.MatchField, value uint64, rng *Range) {
+	action := openflow13.NewNXActionRegLoad(rng.ToNxRange().ToOfsBits(), field, value)
+	a.actions = append(a.actions, action)
+}
+
+// MoveToLabel is an action to move data into ct_mark.
+func (a *ofCTAction) MoveToLabel(fromName string, fromRng, labelRng *Range) CTAction {
+	a.strActions = append(a.strActions, fmt.Sprintf("move:%s[%d..%d]->%s[%d..%d]", fromName, fromRng[0], fromRng[1], NxmFieldCtLabel, labelRng[0], labelRng[1]))
+	fromField, _ := openflow13.FindFieldHeaderByName(fromName, false)
+	toField, _ := openflow13.FindFieldHeaderByName(NxmFieldCtLabel, false)
+	a.move(fromField, toField, uint16(fromRng.length()), uint16(fromRng[0]), uint16(labelRng[0]))
+	return a
+}
+
+func (a *ofCTAction) move(fromField *openflow13.MatchField, toField *openflow13.MatchField, nBits, fromStart, toStart uint16) {
+	action := openflow13.NewNXActionRegMove(nBits, fromStart, toStart, fromField, toField)
+	a.actions = append(a.actions, action)
+}
+
+// CTDone is an action to return FlowBuilder.
+func (a *ofCTAction) CTDone() FlowBuilder {
+	a.builder.Flow.ConnTrack(a.commit, a.force, &a.ctTable, &a.ctZone, a.actions...)
+	if len(a.strActions) > 0 {
+		a.repr += fmt.Sprintf("exec(%s)", strings.Join(a.strActions, ","))
+	}
+	a.builder.actions = append(a.builder.actions, fmt.Sprintf("ct(%s)", a.repr))
+	return a.builder
+}
+
+func (a *ofFlowAction) SetField(key, value string) FlowBuilder {
+	return a.builder
+}
+
+// SetDstMAC is an action to modify packet destination MAC address to the specified address.
+func (a *ofFlowAction) SetDstMAC(addr net.HardwareAddr) FlowBuilder {
+	a.builder.actions = append(a.builder.actions, fmt.Sprintf("set_field:dl_dst->%s", addr.String()))
+	a.builder.SetMacDa(addr)
```

Comment:
out of curiosity, which one takes effect? It looks like both of them are setting fields.

Reply:
Originally I have a specific slice for action strings, and it is used for method `String()` in Flow.  In current code, I have remve that method from interfaces, so the actions slice is removed too.